src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -166,84 +168,211 @@ private ResultRetriever(RangeIterator operation,
 
-                // If we already got some matching data, but the next primary key belongs to a different partition,
-                // then return what we have collected so far.
-                if (!partitionIterators.isEmpty()
-                    && !operation.peek().loadDeferred().partitionKey().equals(currentKey.partitionKey()))
-                    return merge(partitionIterators);
+        /**
+         * Retrieves the next primary key that belongs to the given partition and is selected by the query controller.
+         * The underlying key iterator is advanced only if the key belongs to the same partition.
+         * <p>
+         * Returns null if:
+         * <ul>
+         *   <li>there are no more keys</li>
+         *   <li>the next key is beyond the upper bound</li>
+         *   <li>the next key belongs to a different partition</li>
+         * </ul>
+         * </p>
+         */
+        private @Nullable PrimaryKey nextKeyInPartition(DecoratedKey partitionKey)
+        {
+            PrimaryKey key;
+            do
+            {
+                if (!operation.hasNext())
+                    return null;
+                if (!operation.peek().partitionKey().equals(partitionKey))
+                    return null;
 
-                currentKey = operation.next().loadDeferred();
+                key = nextKey();
             }
+            while (key != null && !controller.selects(key));
+            return key;
         }
 
-        private UnfilteredRowIterator merge(List<UnfilteredRowIterator> iterators) {
-            return iterators.isEmpty() ? endOfData() : UnfilteredRowIterators.merge(iterators);
+        /**
+         * Gets the next key from the underlying operation.
+         * Returns null if there are no more keys <= lastPrimaryKey.
+         */
+        private @Nullable PrimaryKey nextKey()
+        {
+            if (!operation.hasNext())
+                return null;
+            PrimaryKey key = operation.next();
+            return isWithinUpperBound(key) ? key : null;
+        }
+
+        /**
+         * Returns true if the key is not greater than lastPrimaryKey
+         */
+        private boolean isWithinUpperBound(PrimaryKey key)
+        {
+            return lastPrimaryKey.token().isMinimum() || lastPrimaryKey.compareTo(key) >= 0;
+        }
+
+        /**
+         * Gets the next key range from the underlying range iterator.
+         */
+        private @Nullable AbstractBounds<PartitionPosition> nextKeyRange()
+        {
+            return keyRanges.hasNext() ? keyRanges.next().keyRange() : null;
+        }
+
+        /**
+         * Convenience function to skip to a given token.
+         */
+        private void skipTo(@Nonnull Token token)
+        {
+            operation.skipTo(keyFactory.createTokenOnly(token));
+        }
+
+        /**
+         * Skips to the key that belongs to a different partition than the last key we fetched.
+         */
+        private void skipToNextPartition()
```

Comment:
I have reservations about this method but I can't see a better way to do it. My concern is that we are iterating over every row in the partition to get to the next one. It's a problem but I can't see an immediate way to fix it.

Reply:
We are not really iterating anything. If the calling code uses the previous iterator till the end, we'll be already at the next partition and the loop will exit immediately. This is basically to ensure we skip the remaining data in the current partition if the user of this code doesn't do it. So basically this doesn't add any overhead, and it should be still faster than just iterating the iterator returned by `computeNext` (which not only iterates keys from the index, but also retrieves the data!).

Reply:
Thanks. I always see the worst in every situation and was considering having to iterate a huge partition to get to the next.